### PR TITLE
 Delete unreferenced state groups during purge

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/storage/state.py [hunk 1]
```diff
@@ -1041,6 +1041,56 @@ def _count_state_group_hops_txn(self, txn, state_group):
...
+        # descendants of `state_group`
+        state_groups_to_search = [state_group]
+
+        # Set of state groups we've already checked
+        state_groups_searched = set()
+
+        while state_groups_to_search:
+            state_group = state_groups_to_search.pop()  # Next state group to check
+
+            is_referenced = self._simple_select_one_onecol_txn(
```
Comment (richvdh) [inline synapse/storage/state.py]: the problem here is that we might be about to delete a bunch of the events :/. I think you need a `WHERE event_id NOT IN events_to_delete`.

File: synapse/storage/state.py [hunk 2]
```diff
@@ -1041,6 +1041,56 @@ def _count_state_group_hops_txn(self, txn, state_group):
...
+
+        # State groups we need to check, contains state groups that are
+        # descendants of `state_group`
+        state_groups_to_search = [state_group]
+
+        # Set of state groups we've already checked
+        state_groups_searched = set()
+
+        while state_groups_to_search:
+            state_group = state_groups_to_search.pop()  # Next state group to check
```
Comment (richvdh) [inline synapse/storage/state.py]: can we not do this for more than one state group at a time?

File: synapse/storage/state.py [hunk 3]
```diff
@@ -1082,6 +1133,12 @@ def __init__(self, db_conn, hs):
             columns=["state_key"],
             where_clause="type='m.room.member'",
         )
+        self.register_background_index_update(
+            self.EVENT_STATE_GROUP_INDEX_UPDATE_NAME,
+            index_name="event_to_state_groups_sg_index",
```
Comment (richvdh) [inline synapse/storage/state.py]: worth remembering that we already have this on matrix.org, so would ideally not spend ages building another copy of the index.

File: synapse/storage/events.py [hunk 4]
```diff
@@ -2037,13 +2038,33 @@ def _purge_history_txn(
...
-            )
+            for row in rows:
+                sg = row["state_group"]
+
+                if sg in state_groups_to_delete:
+                    # exclude state groups we are about to delete: no point in
+                    # updating them
+                    continue
+
+                if not self._is_state_group_referenced(txn, sg):
```
Comment (richvdh) [inline synapse/storage/events.py]: if you're going to reference methods in StateGroupWorkerStore, can you make EventsStore inherit from it?

Comment (erikjohnston) [reply synapse/storage/events.py]: Oh, yes. I just assumed we already were (since we use functions from there elsewhere I think).

**Comment by erikjohnston** [comment]
I've shuffled things about a bit to checking of unreferenced state groups in batches, rather than one at a time, as well as simplifying the initial query.

File: synapse/storage/state.py [hunk 5]
```diff
@@ -1041,6 +1041,86 @@ def _count_state_group_hops_txn(self, txn, state_group):
...
+        next_to_search = set(state_groups)
+        while next_to_search:
+            # We bound size of groups we're looking up at once, to stop the
+            # SQL query getting too big
+            if len(next_to_search) < 100:
+                current_search = next_to_search
+                next_to_search = set()
+            else:
+                lst = list(next_to_search)
+                current_search = set(lst[:100])
```
Comment (richvdh) [inline synapse/storage/state.py]: I'd be inclined to do:

```
current_search = set(islice(next_to_search, 100))
next_to_search -= current_search
```

File: synapse/storage/state.py [hunk 6]
```diff
@@ -1041,6 +1041,86 @@ def _count_state_group_hops_txn(self, txn, state_group):
...
+            # Check if state groups are referenced
+            sql = """
+                SELECT state_group, count(*) FROM event_to_state_groups
+                LEFT JOIN events_to_purge AS ep USING (event_id)
+                WHERE state_group IN (%s) AND ep.event_id IS NULL
+                GROUP BY state_group
+            """ % (",".join("?" for _ in current_search),)
+            txn.execute(sql, list(current_search))
+
+            referenced = set(sg for sg, cnt in txn if cnt > 0)
```
Comment (richvdh) [inline synapse/storage/state.py]: surely `cnt` will never be zero?

File: synapse/storage/state.py [hunk 7]
```diff
@@ -1041,6 +1041,86 @@ def _count_state_group_hops_txn(self, txn, state_group):
...
+                table="state_group_edges",
+                column="prev_state_group",
+                iterable=current_search,
+                keyvalues={},
+                retcols=("prev_state_group", "state_group",),
+            )
+
+            next_to_search.update(row["state_group"] for row in rows)
+            # We don't bother re-handling groups we've already seen
+            next_to_search -= state_groups_seen
```
Comment (richvdh) [inline synapse/storage/state.py]: should we not do this before adding to next_to_search?

```
prevs = set(row["state_group"] for row in rows)
prevs -= state_groups_seen
next_to_search |= prevs
state_groups_seen |= prevs

File: synapse/storage/state.py (hunk 6 above)
Comment (richvdh) [reply synapse/storage/state.py]: (so why not do `SELECT DISTINCT state_group` rather than the `GROUP BY`)

**Comment by rubo77** [comment]
Is this also deleting leftovers of the old Nuke Script?

**Pull request merged**